107 improve data loading

andkaha opened the pull request:
<!-- Remove sections from this template that are not used -->
## Related issue(s) and PR(s)

This PR closes #107 

<!-- Include below a description of the changes proposed in the pull request -->

## Type of change

<!-- Please delete options that are not relevant -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

## List of changes made

This PR streamlines the process of loading new data into a deployed service. It does this by introducing a loop in `start-script.sh`, which checks every minute to see if new data is available.  If new data is available, the data import scripts are executed to create a new database, which then replaces the existing database.  Once the data import is complete, the raw data is removed, and the service is restarted.

One consequence of this development is the removal of the separate `database` container and its `Dockerfile`.

This PR also combines the `start-script.sh` scripts of the development and production environments into a single script.  The desired functionality depends on how the script is invoked, which is done from the `app` container's `Dockerfile`.

The PR also contains code to clean up the names of some of the compounds by removing the superfluous `[class: ...]` at the end of some names of non-metals.

## Screenshot of the fix

<!-- Attach screenshot if relevant -->

## Testing

Start the service.  It should not matter if the Docker volume is empty or not.

```shell
docker compose up --build
```

Upload the `data-import` directory of raw data into the running container.

```shell
docker compose cp data-import app:/home/bacmet/data/
```

Within a minute, the service should notice the new directory of data and start importing it.

After the import, the data should not be imported again by the service in its 60s loop.  Uploading the data again should also not trigger an import, unless any of the files or directories have had their modification timestamp altered.

If a timestamp modification is found, the contents of the files are verified against the MD5 checksums computed from the raw data after the previous import.  If verification fails, the import is performed; otherwise, it is skipped.

To test re-importing the data, upload it again as above (no re-import should happen, and a minute should pass).

Then `touch` one of the files:

```shell
touch data-import/1-Experimentally_validated/1-BacMet_experimentally_validated_database_3.0.csv
```

Copying the data again should trigger a checksum verification within a minute, and the import should be skipped.

Changing any of the files, e.g., the `data-import/1-Experimentally_validated/1-BacMet_experimentally_validated_database_3.0.csv` file mentioned above, and copying all data into the container again, should trigger an import within a minute.  As a test, you may modify a single letter or word on the first line of that CSV file (the first line is not used by the import script).

(Note that the act of copying the data into the volume may trigger a checksum verification if the upload overlaps with the test for updated files.  This is due to the timestamps of the files that are "in flight" will temporarily be the current time until the copying process is done.)

## Further comments

<!-- Specify questions or related information -->

## Definition of Done checklist

- [x] I have made an effort making the commit history understandable
- [x] I have performed a self-review of my own code and commented any hard-to-understand areas
- [ ] Existing tests are passing and I wrote unit tests for new functionality
- [x] My changes generate no new warnings
- [ ] Make sure the layout and text follow design sketches, if there are any

andkaha committed:
feat: Smarter caching

Use "id" with the "--mount" option to "RUN" to be able to "share"
caches.  We don't see a significant speedup from this, but still...

andkaha committed:
feat: Merge the start scirpt for the app

The development service is invoked by passing "dev" as the 1st argument.

andkaha committed:
fix: Strip off the "[class: ...]" string at the end compound names

andkaha committed:
feat: Remove separate "database" container

Data loading is instead done by copying the data-to-be-imported into the
running container.  The data is loaded when it is seen, then deleted and
the service restarted.

andkaha committed:
feat(docs): Update documentation about data loading too

andkaha committed:
fix: Tabs for indentation

andkaha committed:
fix: Spelling

pdmnyberg commented on app/scripts/start-script.sh:
```diff
@@ -10,4 +10,69 @@ install -d "$XDG_CONFIG_HOME"
+		# data.
+		DATABASE=$DATABASE.new db-scripts/scripts/import-all.sh
+		mv "$DATABASE.new" "$DATABASE"
+		rm -rf "$IMPORT_DIR"
```
I am not too excited about removing the data used for import. I know that in the correct cases this should just work and be a replica of the source data but I always feel a bit uneasy when a script deletes the sources when done. 

Could you help calm my nerves? (What I am afraid of here is that someone would loose data by using the system wrong)

I would like to suggest another solution here though but feel free to opt out if it is too complicated:
- Keep `cache files` for each file containing the hash of the file
- When doing the update check:
  - If timestamp of files > timestamp of `cache files` then continue else abort
  - Calculate new hash for files
  - If hash is mismatch then run the import and update the related `cache files` else update timestamp of `cache files`

I guess one could also opt for using a Makefile to achieve something similar...

andkaha replied:
Is it that you are afraid that someone may try to run `start-script.sh` outside of the Docker container? I will add a simple MD5 check and refrain from deleting the data from the volume.

andkaha replied:
Code is updated. Also, testing instructions written.

andkaha committed:
fix: Don't delete raw data, be smarter with when to do the import

This means keeping the raw data around, and relying on timestamps and
MD5 checksums of the data.

andkaha commented on app/scripts/start-script.sh:
```diff
@@ -10,4 +10,81 @@ install -d "$XDG_CONFIG_HOME"
+		fi
+	fi
+
+	if "${do_import:-false}"; then
```
`true` and `false` are standard utilities. We have full control over the variable `do_import`. It is always either `true`, or it is unset.

andkaha committed:
fix: Grammar

andkaha committed:
fix: Smarter way to find updated files

pdmnyberg commented on app/scripts/start-script.sh:
```diff
@@ -10,4 +10,82 @@ install -d "$XDG_CONFIG_HOME"
+		echo 'Database file does not exist, importing data.' >&2
+		do_import=true
+	elif [ -d "$IMPORT_DIR" ]; then
+                # If the import directory exists, we need to check if
```
The indentation seems off here. You seem to be using spaces instead of tabs as you seem to be using on most other rows?

andkaha replied:
Yes, the indentation is off. I will fix it.

pdmnyberg reviewed (commented):
@andkaha I am curios as to what would happen if an import check is triggered while someone is still uploading files. What happens if there is only a partial upload/update of the data set.

I assume that would cause the import script to run once for the incomplete dataset and possibly fail and once the upload is finished it will run again and succeed. Would that be a correct assumption?

At KB they used an empty file with a specific name to signify that an upload to the inbox was complete. This can be quite finicky to remember so I am not necessarily suggestion that we do that, but do you have any ideas of how we could guard against loading incomplete data?

If you my assumption above is correct that the complete data set should eventually be loaded then I think this is good enough to merge. Nice work!

andkaha commented:
> @andkaha I am curios as to what would happen if an import check is triggered while someone is still uploading files. What happens if there is only a partial upload/update of the data set.
> 
> I assume that would cause the import script to run once for the incomplete dataset and possibly fail and once the upload is finished it will run again and succeed. Would that be a correct assumption?
> 
> At KB they used an empty file with a specific name to signify that an upload to the inbox was complete. This can be quite finicky to remember so I am not necessarily suggestion that we do that, but do you have any ideas of how we could guard against loading incomplete data?
> 
> If you my assumption above is correct that the complete data set should eventually be loaded then I think this is good enough to merge. Nice work!

This is what the 30s sleep is for.  That sleep should hopefully be enough for the upload to finish, even if the new data was detected "mid-upload". Increasing it to 120 seconds would allow for the upload to finish even if it was slow (1 GB over 100 MBit/s takes about 90 seconds). 

Another solution would be to allow the user to upload a "trigger" file that tells the script to load the data.  With such a trigger, there is no need to keep track of checksums. However, this requires extra attention from the service administrator.

What do you think?

andkaha committed:
fix: Sort out tab indentation

pdmnyberg commented:
> > @andkaha I am curios as to what would happen if an import check is triggered while someone is still uploading files. What happens if there is only a partial upload/update of the data set.
> > I assume that would cause the import script to run once for the incomplete dataset and possibly fail and once the upload is finished it will run again and succeed. Would that be a correct assumption?
> > At KB they used an empty file with a specific name to signify that an upload to the inbox was complete. This can be quite finicky to remember so I am not necessarily suggestion that we do that, but do you have any ideas of how we could guard against loading incomplete data?
> > If you my assumption above is correct that the complete data set should eventually be loaded then I think this is good enough to merge. Nice work!
> 
> This is what the 30s sleep is for. That sleep should hopefully be enough for the upload to finish, even if the new data was detected "mid-upload". Increasing it to 120 seconds would allow for the upload to finish even if it was slow (1 GB over 100 MBit/s takes about 90 seconds).
> 
> Another solution would be to allow the user to upload a "trigger" file that tells the script to load the data. With such a trigger, there is no need to keep track of checksums. However, this requires extra attention from the service administrator.
> 
> What do you think?

The "trigger" file would be similar to what I mentioned right? I think that would be an ok choice but I agree with you that is an extra step that can be easy to forget.  I am fine with the current solution if the database will eventually be correct even if the sleep is not enough. Do you also think that the data will eventually be loaded correctly given the current solution?

andkaha committed:
fix: Ensure that the data before import is the same as the data after import

... or else try importing again

andkaha commented:
@pdmnyberg I've implemented the solution with before/after checksumming that we discussed.

pdmnyberg reviewed (approved):
My Bash skills are a bit weak but this looks great to me. The way it works looks reliable and reasonable to me.

Good work!
